sdk/program/src/message/account_keys.rs

```diff
@@ -82,35 +82,45 @@ impl<'a> AccountKeys<'a> {
 
     /// Compile instructions using the order of account keys to determine
     /// compiled instruction account indexes.
-    pub fn compile_instructions(&self, instructions: &[Instruction]) -> Vec<CompiledInstruction> {
-        let account_index_map: BTreeMap<&Pubkey, u8> = BTreeMap::from_iter(
-            self.iter()
-                .enumerate()
-                .map(|(index, key)| (key, index as u8)),
-        );
+    pub fn try_compile_instructions(
+        &self,
+        instructions: &[Instruction],
+    ) -> Result<Vec<CompiledInstruction>, CompileError> {
+        let mut account_index_map = BTreeMap::<&Pubkey, u8>::new();
+        for (index, key) in self.iter().enumerate() {
+            let index = u8::try_from(index).map_err(|_| CompileError::TooManyAccountKeys)?;
```

Comment:
There's a runtime limit of 64 but this compile error is for correctness not for runtime validity. I just want to make sure that the index actually fits in a `u8` because that's how we serialize it in a transaction. There's many ways to create an invalid transaction on the client side so I think it's ok for now to allow large `AccountKeys` to be created because they will be rejected during sanitization and the runtime check could change in the future so I don't want to hardcode that into the client.

Reply:
hmm, the error seems misleading in that case since we can have too many account keys without getting a `TooManyAccountKeys` :thinking:

i typically prefer to fail asap rather than let pathology fester until some other process notices later. killing it at the client, especially in the first implementation, which the others will copy seems like the less wrong thing to do

Reply:
> hmm, the error seems misleading in that case since we can have too many account keys without getting `TooManyAccountKeys` 🤔

Agreed, that was a bad name. I updated to `AccountIndexOverflow` and `AddressTableLookupIndexOverflow`.

> i typically prefer to fail asap rather than let pathology fester until some other process notices later. killing it at the client, especially in the first implementation, which the others will copy seems like the less wrong thing to do

I'm hesitant to do runtime checks on the client because runtime rules can change over time. For example, the serialized size of a transaction could change, the total number of accounts lockable by a transaction (currently 64) could be updated later, etc.